lib/models.py

```diff
@@ -2,13 +2,267 @@
+    verbose: bool = False            # weather or not to print out the CoT
+    selection_method: str = 'greedy' # Describes how set of best possible solutions is selected. Can either be 'greedy' or 'sample'.
+    max_num_tries: int = 3           # maximum number of tries to get a score or final answer before it gives up.
+    return_boolean: bool = True      # flag which, when set to true, forces CoT to output a boolean string (either "true" or "false")
+    get_samples_prompt: str = ''     # prompt to get the samples. Requires "Problem" and "Previous_CoT" as placeholders
+    get_scores_prompt: str = ''      # prompt to get the scores. Requiress "CoT" as a placeholder.
+    get_answer_prompt: str = ''      # prompt to get the final answer. Requires "CoT" and "Problem" as a placeholder.
+
+    def __init__(self, llm, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+        
+        self.llm = llm
+
+        # Check if the neccessary prompts have been loaded into the config
+        if len(self.get_samples_prompt) == 0:
+            raise AssertionError('Sample prompt must be included in CoT configuration')
+        if len(self.get_scores_prompt) == 0:
+            raise AssertionError('Score prompt must be included in CoT configuration')
+        if len(self.get_answer_prompt) == 0:
+            raise AssertionError('Answer prompt must be included in CoT configuration')
+
+    def get_samples(self, problem, previous_CoT):
+        """
+        Takes an input problem, x, and an existing chain of thought, y, 
+        and outputs a continued/enchanced chain of thought.
+
+        Note that y may simply be an empty string, especially if this is
+        in the first layer of the CoT. 
+        """
+        samples = []
+        for _ in range(self.num_generate_samples):
+            prompt = self.get_samples_prompt.format(
+                    Problem=problem, 
+                    Previous_CoT=previous_CoT,
+            )
+            samples.append(self.llm(prompt = prompt))
+        return samples
+
+    def get_score(self, problem, CoT):
+        """
+        Takes as input a problem, x, and a chain of thought about that
+        problem, y, and returns an integer to rate how likely the chain
+        of thought is able to solve the problem.
+
+        1 means that the thougths on correctness are fundamentally incorrect, 
+        5 means that the thoughts on correctness are correct but do not reach a conclusion
+        10 means the thoughts on correctness reach a solid conclusion.
+        
+        Note that this method returns 0 if it cannot get a proper integer from the llm
+        """
+        score = None
+        num_tries = 0
+        while score is None and num_tries < self.max_num_tries:
+
+            prompt = self.get_scores_prompt.format(
+                        CoT = CoT,
```

Comment:
I'd name this `current_CoT` so that you don't overwrite a class named `CoT` that you might inherit later.